### ISSUE TITLE
type deduction fail with Callable[[Literal], _]

### ISSUE THREAD

**Issue by pinterior** [microsoft/pyright#4279]
**Describe the bug**

error reported when function object takes literal parameter and literal values are passed to generic function.

**To Reproduce**

```python3
from typing import Callable, Literal, Tuple, TypeVar
from typing_extensions import TypeVarTuple, Unpack

T = TypeVar("T")

def p0(v: T, f: Callable[[T], None]): ...
def t0(v: Literal[1, 2], f: Callable[[Literal[1, 2]], None]):
    p0(v, f)

Ts = TypeVarTuple("Ts")

def p1(f: Callable[[Unpack[Ts]], None], vs: Tuple[Unpack[Ts]]): ...
def t1(f: Callable[[Literal[1, 2]], None], vs: Tuple[Literal[1, 2]]):
    p1(f, vs)
```
```
q.py:8:11 - error: Argument of type "(Literal[1, 2]) -> None" cannot be assigned to parameter "f" of type "(T@p0) -> None" in function "p0"
    Type "(Literal[1, 2]) -> None" cannot be assigned to type "(int) -> None"
      Parameter 1: type "int" cannot be assigned to type "Literal[1, 2]"
        Type "int" cannot be assigned to type "Literal[1, 2]"
          "int" cannot be assigned to type "Literal[1]"
          "int" cannot be assigned to type "Literal[2]" (reportGeneralTypeIssues)
q.py:14:11 - error: Argument of type "Tuple[Literal[1, 2]]" cannot be assigned to parameter "vs" of type "Tuple[*Ts@p1]" in function "p1"
    "Tuple[Literal[1, 2]]" is incompatible with "Tuple[*Ts@p1]"
      Tuple entry 1 is incorrect type
        Type "*tuple[int]" cannot be assigned to type "*tuple[Literal[1, 2]]"
          "*tuple[int]" is incompatible with "*tuple[Literal[1, 2]]"
            Tuple entry 1 is incorrect type (reportGeneralTypeIssues)
```

**Expected behavior**

no error reported

**VS Code extension or command-line**

command-line tool, 1.1.282

**Comment by erictraut** [comment]
Pyright's constraint solver contains a bunch of heuristics about whether to retain literal values. In most cases, it doesn't. That's normally the correct answer.

```python
def foo(x: T) -> T: ...

reveal_type(foo(1)) # You would expect `int` here rather than `Literal[1]`
```

In this case, the heuristics are failing. I don't know if I can fix them without breaking a bunch of common cases. I'll need to investigate further. The fact that this type checks in mypy leads me to believe there is probably a viable solution.

**Comment by erictraut** [comment]
I've made a pretty significant change to the constraint solver logic so it handles better (and more consistent) handling of literals when solving for a TypeVar. This fixes the first issue in this bug report. It does not yet address the second issue related to TypeVarTuples. This will require additional work.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.286, which I just published. It will also be included in a future version of pylance.
